### ISSUE TITLE
Ember-CLI and Ember Guides disagree about helper registration

### ISSUE THREAD

**Issue by jamesarosen** [ember-cli/ember-cli#3710]
The [v1.11.0 guides](http://guides.emberjs.com/v1.11.0/templates/writing-helpers/) suggest
```js
// app/helpers/my-helper.js
export default Ember.Handlebars.makeBoundHelper(function(x, y) { ... });
```
but ember-cli's [v0.2.2 helper blueprint](https://github.com/ember-cli/ember-cli/blob/v0.2.2/blueprints/helper/files/app/helpers/__name__.js#L7) generates
```js
export myHelper(x, y) {...}

export default Ember.HTMLBars.makeBoundHelper(myHelper);
```
Note specifically the `Ember.Handlebars.makeBoundHelper` vs `Ember.HTMLBars.makeBoundHelper`. These are not identical.

On Ember 1.11.0 and ember-cli-htmlbars 0.7.4, I was using `Ember.HTMLBars.makeBoundHelper` but found it broken. Specifically, my helper was receiving not the individual arguments `x` and `y`, but an Array containing the values for `x` and `y`. Changing this to `Ember.Handlebars.makeBoundHelper` fixes the problem.

In summary, I see two problems:
 1. `Ember.HTMLBars.makeBoundHelper` doesn't work for me. That's not a problem for this project.
 2. In trying to resolve (1), I found an inconsistency between this project and the Ember guides. They should agree.

**Comment by rwjblue** [comment]
The HTMLBars helper syntax is different (as documented at http://emberjs.com/api/classes/Ember.HTMLBars.html#method_makeBoundHelper).  It removes the need to constantly shift arguments in an older style helpers.

Both `Ember.HTMLBars.makeBoundHelper` and `Ember.Handlebars.makeBoundHelper` work, I prefer the new syntax (and to avoid the crazy arg slicing).

**Comment by jamesarosen** [comment]
I do see the difference. I hadn't noticed it before.

I guess I'm just down to problem 1: the Helper file that Ember-CLI generates doesn't work. For some reason, the two versions seem to be *switched* in my app.

**Comment by rwjblue** [comment]
>  the Helper file that Ember-CLI generates doesn't work

Can you describe what you mean by this?  Possibly throw up a demo repo?

**Comment by jamesarosen** [comment]
> Can you describe what you mean by [the helper doesn't work]?

I've expanded [in a gist](https://gist.github.com/jamesarosen/174f8a953b1466aa8981)

> Possibly throw up a demo repo?

I'll work on that tomorrow.

**Comment by jamesarosen** [comment]
I just realized that what I describe in the gist is *completely consistent* with the API docs in http://emberjs.com/api/classes/Ember.HTMLBars.html#method_makeBoundHelper. Those docs are private, however, and there's no guide for using `Ember.HTMLBars.makeBoundHelper`.

Summary: I had no idea that the standard ("standard" as suggested by `ember g helper`) API for helpers was now `helper(array, hash)`.

Does the output of `ember g helper` suggest that `Ember.HTMLBars.makeBoundHelper` should be public?

**Comment by rwjblue** [comment]
@jamesarosen - Great question! I was under the impression that it was already public, I will confirm with others....

**Comment by sheriffderek** [comment]
@jamesarosen @rwjblue - is there something I can do to get in the club where I can see some content at that link you referenced? I'm finding the handle -> html bars switch a little mysterious in places like this.

**Comment by jamesarosen** [comment]
@sheriffderek click the "private" box :)

**Comment by jamesarosen** [comment]
@rwjblue since `makeBoundHelper` is currently private, we have an opportunity to change its behavior before making it public! Is there *any* way to make the helper signature the same as what gets passed in the template? That is, if my template uses the helper like so:
```hbs
{{my-helper binding.to.some.date "a static string" an-option=binding.to.a.string}}
```
then I'd *really* like to have my helper's signature
```js
function myHelper(date, string, options) {
}
```
Without that, I need to go through some hoops to use the same function as a helper and in JS code. Either I need to write a version of the helper that accepts the arguments I want and then wrap it in an HTMLBars-friendly helper function:
```js
export function myHelper(date, string, options) {
}

export default Ember.HTMLBars.makeBoundHelper(function(params, options) {
  const args = params.slice();
  args.push(options);
  return myHelper.apply(this, args);
}
```
or I need to munge my parameters in JS code that uses the helper as a pure function:
```js
import { myHelper } from "app/helpers/my-helper";
//...
return myHelper([ date, string ], options);
```

**Comment by jamesarosen** [comment]
For now, I wrote
```js
// app/utils/make-bound-helper.js
export default function makeBoundHelper(fn) {
  return Ember.HTMLBars.makeBoundHelper(function(ordered, hash) {
    const args = ordered.slice();
    args.push(hash);
    return fn.apply(this, args);
  });
}
```
That gives me the API I expect and isolates the use of private functions to one file :)

**Comment by rwjblue** [comment]
Being forced to munge `arguments` to get the actual "parameters" split from the `options` is very annoying and generally deopts the entire function if done incorrectly.  The `HTMLBars` version is much easier to deal with varying argument lengths.

Ultimately, we are very unlikely to change this since helpers in glimmer are extremely different (*again*).  Doesn't make sense to force folks to change their existing `Ember.HTMLBars.makeBoundHelper` usage and make it public, just to immediately deprecate (in 1.13).

**Comment by jamesarosen** [comment]
Is there somewhere I can see an example of helper registration in Glimmer?

**Comment by jamesarosen** [comment]
> Being forced to munge arguments to get the actual "parameters" split from the options is very annoying and generally deopts the entire function if done incorrectly.

It seems about as difficult as munging to get all the ordered parameters out of the Array. I guess ES6 destructuring makes ordered-slicing a bit easier:

```js
function formatDate(ordered, hash) {
  [ date, format ] = ordered;
  ...
}
```

**Comment by rwjblue** [comment]
Here is the [each](https://github.com/emberjs/ember.js/blob/idempotent-rerender/packages/ember-htmlbars/lib/helpers/each.js).  It does still use similar semantics to `Ember.HTMLBars.makeBoundHelper` (two arguments: `orderedParams` and `hashParams`).

> It seems about as difficult as munging to get all the ordered parameters out of the Array

Perhaps, **but** it does not cause your function to deopt...

> I guess ES6 destructuring makes ordered-slicing a bit easier

Exactly.

**Comment by jamesarosen** [comment]
The guides still use `Ember.Handlebars.makeBoundHelper`, which is not documented in the API docs at all. Ember-CLI still generates the `HTMLBars` version, which is private.

We need *some* public, documented helper registration method.

**Comment by jamesarosen** [comment]
Perhaps we can clear this up in emberjs/rfcs#53
